### PR TITLE
[Woo POS] Products - Pull to refresh

### LINKED ISSUE #11755: [Woo POS] Product list: Pull to refresh
Ref: p1718853120102309/1718762740.351459-slack-C070SJRA8DP

> Can we get a pull-to-refresh on the products list?

### PR DESCRIPTION
**Pull request by samiuelson**
Closes: #11755 

### Description

This PR adds pull to refresh feature to the Products screen in Woo POS. The refresh function resets the pagination to the first page (25 items) of products and triggers (soft) sync of products by sending a request to fetch products from endpoint. After receiving response the update or insert operation is performed and new products are added to the product list.

**Note**: the pull to refresh doesn't remove stale (remotely deleted products). We decided not to wipe product table during pull to refresh because it's used in the store management app. We can consider implementing in-memory storage for Woo POS product selector, independent from store management mode. This would allow implementing the full sync. ([issue](https://github.com/orgs/woocommerce/projects/247/views/3?pane=issue&itemId=69809424))

### Steps to reproduce

Add a new product and verify it's present in the product selector after pulling to refresh.

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/4527432/b1f46c04-2cc9-47eb-9fb6-2f94279bcddf

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt [hunk 1]
```diff
@@ -49,6 +49,19 @@ class WooPosProductsViewModel @Inject constructor(
...
                 onItemClicked(event.item)
             }
+
+            WooPosProductsUIEvent.PullToRefreshTriggered -> {
+                reloadProducts()
+            }
+        }
+    }
+
+    private fun reloadProducts() {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:62]: np: what about 

```
    private fun reloadProducts() {
        viewModelScope.launch {
            val currentState = viewState.value
            if (currentState !is WooPosProductsViewState.Content) return@launch
            _viewState.value = currentState.copy(refreshingProducts = true)
            productsDataSource.loadSimpleProducts()
            _viewState.value = currentState.copy(refreshingProducts = false)
        }
    }
```
?

**Comment by kidinov** [comment]
There is a weird lag of the pull to refresh indicator when it goes up. Any clue what this may happen?


https://github.com/woocommerce/woocommerce-android/assets/4923871/2973f136-2f10-4d1e-a5b5-fcf76f2a1dfc

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:62]: Also, maybe a test?

**Comment by kidinov** [comment]
> the pull to refresh doesn't remove stale (remotely deleted products). 

On the products list in the app, when I move a product to the bin on the web and then either pull to refresh or just open the screen - the deleted product is gone after request comes back. Do they drop the table or how they solve that?

In the POS, I can not make the product be removed at all (PTR or not), which, in my opinion, is a problem we should solve

Also, I am wondering what will happen if we try to create an order with a deleted product? The endpoint that creates an order will return an error?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt [hunk 2]
```diff
@@ -59,15 +66,18 @@ fun WooPosProductsScreen(modifier: Modifier = Modifier) {
         productsStateFlow = productsViewModel.viewState,
         onItemClicked = { productsViewModel.onUIEvent(ItemClicked(it)) },
         onEndOfProductsGridReached = { productsViewModel.onUIEvent(EndOfProductsGridReached) },
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt]: np: seems that we forgot to rename both the function and the event from Grid to List =)

**Review by kidinov — COMMENTED** [review]
Looks good! Thanks. I left a couple of comments, please take a look

**Comment by samiuelson** [comment]
> On the products list in the app, when I move a product to the bin on the web and then either pull to refresh or just open the screen - the deleted product is gone after request comes back. Do they drop the table or how they solve that?

There is no pull to refresh in products selector in the store management mode.

When pull to refresh is triggered on the Products screen, the products table is wiped (in the `ProductListRepository` the `WCProductStore.fetchProducts` is called with default arg `forceRefresh: Boolean = true`). In the case of fetching products when opening the Product selector `WCProductStore.fetchProducts` is called in `ProductSelectorRepository` with `forceRefresh: Boolean = false`.

> In the POS, I can not make the product be removed at all (PTR or not), which, in my opinion, is a problem we should solve
> Also, I am wondering what will happen if we try to create an order with a deleted product? The endpoint that creates an order will return an error?

The endpoint creates the order correctly with products that are deleted but still in trash. In case they are deleted from trash, the order will be created successfully, but with blank items (totals and subtotals = 0).

**Comment by kidinov** [comment]
@samiuelson 👋
 
> There is no pull to refresh in products selector in the store management mode.

I was talking about the product list:

https://github.com/woocommerce/woocommerce-android/assets/4923871/5006aafc-03c7-4fcc-a063-5b6647585b8d

> When pull to refresh is triggered on the Products screen, the products table is wiped (in the ProductListRepository the WCProductStore.fetchProducts is called with default arg forceRefresh: Boolean = true). In the case of fetching products when opening the Product selector WCProductStore.fetchProducts is called in ProductSelectorRepository with forceRefresh: Boolean = false.

Shell, do we do the same, or do you see some downsides to this approach?

> The endpoint creates the order correctly with products that are deleted but still in trash. In case they are deleted from trash, the order will be created successfully, but with blank items (totals and subtotals = 0).

👍

**Comment by samiuelson** [comment]
👋🏼 @kidinov. After looking into how the pull to refresh works in the Products screen, I decided to implement full sync in the Product selector in the POS. Let me know what you think!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt (hunk 2 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt]: [118a94f](https://github.com/woocommerce/woocommerce-android/pull/11901/commits/118a94f2913b92ee0280130fb22e1cf9d1c234ba)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt (hunk 1 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt:62]: Added tests [118a94f](https://github.com/woocommerce/woocommerce-android/pull/11901/commits/118a94f2913b92ee0280130fb22e1cf9d1c234ba), https://github.com/woocommerce/woocommerce-android/pull/11901/commits/35cc138d2eb03d02cac51e348125f37b81496697, and simplified the code bec2bfcb51446f2e541c8557782e208750610fe0

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt [hunk 3]
```diff
@@ -10,17 +11,25 @@ import javax.inject.Inject
...
+    private val site: SelectedSite,
+) {
+    val hasMorePages = handler.canLoadMore.get()
 
     val products: Flow<List<Product>> = handler.productsFlow
         .map { it.filter { product -> product.price != null } }
 
-    suspend fun loadSimpleProducts(): Result<Unit> =
-        handler.loadFromCacheAndFetch(
+    suspend fun loadSimpleProducts(forceRefreshProducts: Boolean = false): Result<Unit> {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: np: I would recommend not having a default value here. This is really not obvious that by default `loadSimpleProducts` is not returning you an actual snapshot of the products, but the cache

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt [hunk 4]
```diff
@@ -10,17 +11,25 @@ import javax.inject.Inject
...
+        if (forceRefreshProducts) {
+            productStore.deleteProductsForSite(site.getOrNull()!!)
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: np: may to use `site.get()`, which returns a non-nullable value already and throws a more appropriate `SelectedSiteException` expression?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt [hunk 5]
```diff
@@ -102,37 +112,54 @@ private fun WooPosProductsScreen(
...
+                }
+            }
             item {
-                ProductLoadingItem()
+                Spacer(modifier = Modifier.height(16.dp))
             }
         }
-        item {
-            Spacer(modifier = Modifier.height(16.dp))
+        PullRefreshIndicator(
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt]: This is part of the "loaded" state, but now, when we are reloading products we always switch the "loading" state so the indicator is gone, right after it's pulled down

The best would be to show the list and PTR indicator before the data comes back but with the implementation of the `WCProductStore` this maybe not really easy to do, as we can not remove and update the cache update successfully loaded the first page of products from the server? Probably is not the biggest issue, so up to you if you think it worths fixing

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt [hunk 6]
```diff
@@ -10,17 +11,25 @@ import javax.inject.Inject
...
 
 @Singleton
-class WooPosProductsDataSource @Inject constructor(private val handler: ProductListHandler) {
-    var hasMorePages = handler.canLoadMore.get()
+class WooPosProductsDataSource @Inject constructor(
+    private val handler: ProductListHandler,
+    private val productStore: WCProductStore,
+    private val site: SelectedSite,
+) {
+    val hasMorePages = handler.canLoadMore.get()
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: Loading another page doesn't work anymore, I guess because it's val now?

I think the best to have it as prop:

```
val hasMorePages
        get() = handler.canLoadMore.get()
```

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt [hunk 7]
```diff
@@ -45,19 +45,34 @@ class WooPosProductsViewModel @Inject constructor(
...
+        viewModelScope.launch {
+            val currentState = viewState.value as? WooPosProductsViewState.Content ?: return@launch
+            _viewState.value = currentState.copy(refreshingProducts = true)
+            productsDataSource.loadSimpleProducts(forceRefreshProducts = true)
+            _viewState.value = currentState.copy(refreshingProducts = false)
         }
     }
 
     private suspend fun calculateViewState(products: List<Product>): WooPosProductsViewState {
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt]: np: the method is a bit hard to follow now, maybe to simplify:

```
    private suspend fun calculateViewState(products: List<Product>): WooPosProductsViewState =
        when {
            products.isEmpty() && !isRefreshingProducts() -> WooPosProductsViewState.Empty
            products.isEmpty() && isRefreshingProducts() -> WooPosProductsViewState.Loading
            else -> products.toContentState()
        }

    private suspend fun List<Product>.toContentState() = WooPosProductsViewState.Content(
        products = map { product ->
            WooPosProductsListItem(
                id = product.remoteId,
                name = product.name,
                price = priceFormat(product.price),
                imageUrl = product.firstImageUrl,
            )
        },
        loadingMore = false,
        refreshingProducts = false,
    )
```

**Review by kidinov — COMMENTED** [review]
@samiuelson I left a few minor comments, and there is one that must be fixed. Please take a look

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/225">19.5</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt (hunk 7 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsViewModel.kt]: Done - 359d8bf825db3901638aa3eaf0668831c1109d17

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt (hunk 6 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: Good catch, fixed - f85714f314cc13aa6332f423213f8ce782de36d1

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt (hunk 5 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsScreen.kt]: I just moved the `PullRefreshIndicator` up, to allow pulling to refresh in other states (Empty/Error) as well. I think it's good enough for now.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt (hunk 4 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/products/WooPosProductsDataSource.kt]: 19888c83e18ca643de4a743a7119c9894f49d108

**Review by kidinov — APPROVED** [review]
Looks good! There is a nitpick which you can ignore and a merge conflict to resolve
